### PR TITLE
8180352: Add Stream.toList() method

### PR DESCRIPTION
**Pull request by stuart-marks**
This change introduces a new terminal operation on Stream. This looks like a convenience method for Stream.collect(Collectors.toList()) or Stream.collect(Collectors.toUnmodifiableList()), but it's not. Having this method directly on Stream enables it to do what can't easily by done by a Collector. In particular, it allows the stream to deposit results directly into a destination array (even in parallel) and have this array be wrapped in an unmodifiable List without copying.

In the past we've kept most things from the Collections Framework as implementations of Collector, not directly on Stream, whereas only fundamental things (like toArray) appear directly on Stream. This is true of most Collections, but it does seem that List is special. It can be a thin wrapper around an array; it can handle generics better than arrays; and unlike an array, it can be made unmodifiable (shallowly immutable); and it can be value-based. See John Rose's comments in the bug report:

https://bugs.openjdk.java.net/browse/JDK-8180352?focusedCommentId=14133065&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14133065

This operation is null-tolerant, which matches the rest of Streams. This isn't specified, though; a general statement about null handling in Streams is probably warranted at some point.

Finally, this method is indeed quite convenient (if the caller can deal with what this operation returns), as collecting into a List is the most common stream terminal operation.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8180352](https://bugs.openjdk.java.net/browse/JDK-8180352): Add Stream.toList() method

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1026/head:pull/1026`
`$ git checkout pull/1026`

### CODE REVIEW

**Review by dfuch — COMMENTED** [review]
Should there be a test that tests the default implementation in `j.u.s.Stream`? Or maybe there is and I missed that?

File: src/java.base/share/classes/java/util/stream/Stream.java [hunk 1]
```diff
@@ -1162,6 +1162,36 @@ default DoubleStream mapMultiToDouble(BiConsumer<? super T, ? super DoubleConsum
      */
     <R, A> R collect(Collector<? super T, A, R> collector);
 
+    /**
+     * Accumulates the elements of this stream into a {@code List}. The elements in
+     * the list will be in this stream's encounter order, if one exists. There are no
+     * guarantees on the implementation type, mutability, serializability, or
```
Comment (jodastephen) [inline src/java.base/share/classes/java/util/stream/Stream.java:1168]: It would be useful for callers to feel more confident that they will get an immutable instance. In java.time.* we have wording like "This interface places no restrictions on the mutability of implementations, however immutability is strongly recommended." Could something like that work here, emphasising that everyone implementing this method should seek to return an immutable list?

File: test/jdk/java/util/stream/test/org/openjdk/tests/java/util/stream/ToListOpTest.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ */
+@Test
+public class ToListOpTest extends OpTestCase {
+
+    public void testToList() {
+        assertCountSum(countTo(0).stream().toList(), 0, 0);
+        assertCountSum(countTo(10).stream().toList(), 10, 55);
+    }
+
+    private void checkUnmodifiable(List<Integer> list) {
```
Comment (fweimer) [inline test/jdk/java/util/stream/test/org/openjdk/tests/java/util/stream/ToListOpTest.java:47]: I'd expect a test here that if the list contains a null element, `List::copyOf` throws `NullPointerException`.

File: src/java.base/share/classes/java/util/ImmutableCollections.java [hunk 3]
```diff
@@ -160,6 +170,81 @@ private ImmutableCollections() { }
...
+     * @param input the input array
+     * @return the new list
+     */
+    @SuppressWarnings("unchecked")
+    static <E> List<E> listFromTrustedArray(Object... input) {
+        for (Object o : input) {
+            Objects.requireNonNull(o);
+        }
+
+        switch (input.length) { // implicit null check of elements
```
[the remark this answers is not in the document]
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/ImmutableCollections.java:211]: Yes, actually that comment belongs up above. I'll fix it, thanks.

File: src/java.base/share/classes/java/util/stream/Stream.java (hunk 1 above)
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/stream/Stream.java:1168]: Yes, good point, the "no guarantee of mutability" clashes with the later statement about the possibility of the returned instance being value-based, which strongly implies immutability. I'll work on tuning this up to be a stronger statement on immutability, while retaining "no-guarantee" for implementation type, serializability, etc. I think we do want to preserve future implementation flexibility in those areas.

**Comment by stuart-marks** [comment]
@dfuch wrote:
> Should there be a test that tests the default implementation in `j.u.s.Stream`? Or maybe there is and I missed that?

The test method `testDefaultOps` does that. The stream test framework has a thing called `DefaultMethodStreams` that delegates everything except default methods to another Stream instance, so this should test the default implementation.

**Comment by stuart-marks** [comment]
OK, there are rather too many forking threads here for me to try to reply to any particular message, so I'll try to summarize things and say what I intend to do.

Null tolerance. While part of me wants to prohibit nulls, the fact is that Streams pass through nulls, and toList() would be much less useful if it were to reject nulls. The affinity here is closer to Stream.toArray(), which allows nulls, rather than Collectors.to[Unmodifiable]List.

Unmodifiability. Unlike with nulls, this is where we've been taking a strong stand recently, where new constructs are unmodifiable ("shallowly immutable"). Consider the Java 9 unmodifiable collections, records, primitive classes, etc. -- they're all unmodifiable. They're data-race-free and are resistant to a whole class of bugs that arise from mutability.

Unfortunately, the combination of the above means that the returned List is neither like an ArrayList nor like an unmodifiable list produced by List.of() or by Collectors.toUnmodifiableList(). [Heavy sigh.] While I've been somewhat reluctant to introduce this new thing, the alternatives of trying to reuse one of the existing things are worse.

John and Rémi pointed out that the way I implemented this, using a subclass, reintroduces the possibility of problems with megamorphic dispatch which we had so carefully avoided by reducing the number of implementation classes in this area. I agree this is a problem.

I also had an off-line discussion with John where we discussed the serialization format, which unfortunately is somewhat coupled with this issue. (Fortunately I think it can mostly be decoupled.) Given that we're introducing a new thing, which is an unmodifiable-list-that-allows-nulls, this needs to be manifested in the serialized form of these new objects. (This corresponds to the new tag value of IMM_LIST_NULLS in the CollSer class.) The alternative is to reuse the existing serialized form, IMM_LIST, for both of these cases, relaxing it to allow nulls. This would be a serialization immutability problem. Suppose I had an application that created a data structure that used lists from List.of(), and I had a global assertion over that structure that it contained no nulls. Further suppose that I serialized and deserizalized this structure. I'd want that assertion to be preserved after deserialization. If another app (or a future version of this app) created the structure using Stream.toList(), this would allow nulls to leak into that structure and violate that assertion. Therefore, the result of Stream.toList() should not be serialization-compatible with List.of() et. al. That's why there's the new IMM_LIST_NULLS tag in the serial format.

However, the new representation in the serial format does NOT necessarily require the implementation to be a different class. I'm going to investigate collapsing ListN and ListNNullsAllowed back into a single class, while preserving the separate serialized forms. This should mitigate the concerns about megamorphic dispatch.

File: src/java.base/share/classes/java/util/stream/Stream.java [hunk 4]
```diff
@@ -1162,6 +1162,36 @@ default DoubleStream mapMultiToDouble(BiConsumer<? super T, ? super DoubleConsum
...
+     * @apiNote If more control over the returned object is required, use
+     * {@link Collectors#toCollection(Supplier)}.
+     *
+     * @return a List containing the stream elements
+     *
+     * @since 16
+     */
+    @SuppressWarnings("unchecked")
+    default List<T> toList() {
+        return (List<T>) Collections.unmodifiableList(new ArrayList<>(Arrays.asList(this.toArray())));
```
Comment (orionll) [inline src/java.base/share/classes/java/util/stream/Stream.java:1192]: Why can't we return `listFromTrustedArrayNullsAllowed` here as in `ReferencePipeline`?
Or at least, we should avoid unnecessary copying of arrays. See [how this is done](https://github.com/amaembo/streamex/blob/master/src/main/java/one/util/streamex/AbstractStreamEx.java#L1313) in StreamEx.

**Review by orionll — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/util/stream/Stream.java (hunk 4 above)
Comment (amaembo) [reply src/java.base/share/classes/java/util/stream/Stream.java:1192]: `listFromTrustedArrayNullsAllowed` is clearly not an option, as it will produce shared-secret leaking (see [JDK-8254090](https://bugs.openjdk.java.net/browse/JDK-8254090) for a similar case). StreamEx solution is dirty as it relies on the implementation detail. I believe, OpenJDK team is not very interested in providing optimal implementations for third-party stream implementations, as third-party implementations will likely update by themselves when necessary. At least, my suggestion to make the default `mapMulti` implementation better was declined.

File: src/java.base/share/classes/java/util/ImmutableCollections.java [hunk 5]
```diff
@@ -160,6 +170,81 @@ private ImmutableCollections() { }
...
+     * <p>Avoids creating a List12 instance, as it cannot accommodate null elements.
+     *
+     * @param <E> the List's element type
+     * @param input the input array
+     * @return the new list
+     */
+    @SuppressWarnings("unchecked")
+    static <E> List<E> listFromTrustedArrayNullsAllowed(Object... input) {
+        if (input.length == 0) {
+            return (List<E>) EMPTY_LIST;
```
Comment (orionll) [inline src/java.base/share/classes/java/util/ImmutableCollections.java:240]: If we return a `ListN` here, does this mean that `Stream.of().toList().contains(null)` will throw an NPE? But this is incorrect because `toList()` returns a null-tolerant List.

File: src/java.base/share/classes/java/util/ImmutableCollections.java [hunk 6]
```diff
@@ -160,6 +170,81 @@ private ImmutableCollections() { }
...
+    }
+
+    /**
+     * Creates a new List from a trusted array, checking for and rejecting null
+     * elements.
+     *
+     * <p>A trusted array has no references retained by the caller. It can therefore be
+     * safely reused as the List's internal storage, avoiding a defensive copy. Declared
+     * with Object... instead of E... as the parameter type so that varargs calls don't
+     * accidentally create an array of type other than Object[].
```
Comment (ePaul) [inline src/java.base/share/classes/java/util/ImmutableCollections.java:199]: Why would that be a problem? If the resulting list is immutable, then the actual array type doesn't really matter, right?

**Comment by ePaul** [comment]
Looking at the linked issue, I see [this comment from Rémi Forax](https://bugs.openjdk.java.net/browse/JDK-8180352?focusedCommentId=14171626&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14171626):

> Please talk with Brian about this change because it nulls a property of the Stream API we (the lambda-util group) have take time to keep.
The whole Stream API doesn't depends on the Collection API, [...] so the Stream API can be easily integrated with other collection API if in the future we want by example add a persistent collection API with no link with java.util.List.

That's an argument I can understand – as is, the Stream API works just as well with collections from e.g. Scala's or Kotlin's (or Ceylon's) collection libraries instead of the java.util ones, just using different collectors. Adding a method which directly returns a java.util.List somewhat couples it to the Java Collection API.

Now this was mentioned two and a half year ago. Did something change which made this consideration irrelevant? I would expect at least some mention of it in the discussion here.

File: src/java.base/share/classes/java/util/ImmutableCollections.java (hunk 5 above)
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/ImmutableCollections.java:240]: Yes, good point, we might need to have a null-tolerant empty list.

File: src/java.base/share/classes/java/util/ImmutableCollections.java (hunk 6 above)
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/ImmutableCollections.java:199]: It's an implementation invariant that the internal array be Object[]. Having it be something other than Object[] can lead to subtle bugs. See [JDK-6260652](https://bugs.openjdk.java.net/browse/JDK-6260652) for example.

File: test/jdk/java/util/stream/test/org/openjdk/tests/java/util/stream/ToListOpTest.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        checkUnmodifiable(objects);
+    }
+
+    @Test(dataProvider = "withNull:StreamTestData<Integer>", dataProviderClass = StreamTestDataProvider.class)
+    public void testDefaultOps(String name, TestData.OfRef<Integer> data) {
+        List<Integer> objects = exerciseTerminalOps(data, s -> DefaultMethodStreams.delegateTo(s).toList());
+        checkUnmodifiable(objects);
+    }
+
+    @Test(dataProvider = "withNull:StreamTestData<Integer>", dataProviderClass = StreamTestDataProvider.class)
```
Comment (PaulSandoz) [inline test/jdk/java/util/stream/test/org/openjdk/tests/java/util/stream/ToListOpTest.java:73]: Given the non-default `toList()` implementation defers to the `toArray()` terminal op there is no need for this and the following tests, which are really designed to shake out the optimizations when producing and operating on arrays.

Comment (stuart-marks) [reply test/jdk/java/util/stream/test/org/openjdk/tests/java/util/stream/ToListOpTest.java:73]: OK, I'll remove all the tests starting from here to the end of the file. I'm assuming that's the set of tests you're referring to.

Comment (PaulSandoz) [reply test/jdk/java/util/stream/test/org/openjdk/tests/java/util/stream/ToListOpTest.java:73]: Yes from line 73 to <SHIFT>G (the end).

**Comment by stuart-marks** [comment]
@ePaul wrote:

> The Stream API works just as well with [third party] collection libraries instead of the java.util ones, just using different collectors. Adding a method which directly returns a java.util.List somewhat couples it to the Java Collection API.
> 
> Now this was mentioned two and a half year ago. Did something change which made this consideration irrelevant? I would expect at least some mention of it in the discussion here.

The separation between streams and the java.util Collections Framework is a good design principle, but it isn't an ironclad rule. It's still easy to have streams create instances of other collections libraries using the Collector interface. What's different here is that the Collections Framework has "leaked into" streams a little bit more, so that they're now more interdependent. This doesn't seem to have any disadvantages; it seems unlikely that the Collections Framework will ever be unplugged from the JDK. However, the benefits are that a List is the closest thing we have to an unmodifiable array that also plays well with generics and that can be value-based; these benefits are considerable.

**Comment by stuart-marks** [comment]
> Simon Roberts wrote:

> This discussion of unmodifiable lists brings me back to the thought that
> there would be good client-side reasons for inserting an UnmodifiableList
> interface as a parent of LIst, not least because all our unmodifiable
> variants from the Collections.unmodifiableList proxy onward fail the Liskov
> substitution test for actually "being contract-fulfilling Lists".

At some point there probably will need to be a long article explaining all the issues here, but at the moment the best writeup I have is this one:

https://stackoverflow.com/a/57926310/1441122

TL;DR there are a few different ways to approach retrofitting something like this, but they all have enough compromises that the net benefits are unclear.

**Comment by plevart** [comment]
Hi Stuart,

I would like to discuss the serialization. You introduce new CollSer.IMM_LIST_NULLS type of immutable collection. This means that if this change goes into JDK16 for example, JDK15 and before will not be able to deserialize such list as they know nothing about IMM_LIST_NULLS even if such lists don't contain nulls. The reason you say to chose new type of serialization format is the following:

"Suppose I had an application that created a data structure that used lists from List.of(), and I had a global assertion over that structure that it contained no nulls. Further suppose that I serialized and deserizalized this structure. I'd want that assertion to be preserved after deserialization. If another app (or a future version of this app) created the structure using Stream.to
 List(), this would allow nulls to leak into that structure and violate that assertion. Therefore, the result of Stream.toList() should not be serialization-compatible with List.of() et. al. That's why there's the new IMM_LIST_NULLS tag in the serial format"

I don't quite get this reasoning. Let's try to decompose the reasoning giving an example. Suppose we had the following data structure:

public class Names implements Serializable {
  private final List<String> names;
  Names(List<String> names) {
    this.names = names;
  }
  public List<String> names() { return names; }
}

App v1 creates such structures using new Names(List.of(...)) and serializes/deserializes them. They keep the invariant that no nulls are present. Now comes App v2 that starts using new Names(stream.toList()) which allows nulls to be present. When such Names instance from app v2 is serialized and then deserialized in app v1, nulls "leak" into data structure of app v1 that does not expect them.

But the question is how does having a separate CollSer.IMM_LIST_NULLS type prevent that from happening?

**Comment by plevart** [comment]
I can see that having a separate IMM_LIST_NULLS type might be necessary to preserve the allows-null/disallows-null behaviour of indexOf and lastIndexOf methods...

NOTE ALSO that while ListN.equals(o) method is using Objects.equals(o1, o2) to compare elements, hashCode is inherited from AbstractImmutableList:

```
        public int hashCode() {
            int hash = 1;
            for (int i = 0, s = size(); i < s; i++) {
                hash = 31 * hash + get(i).hashCode();
            }
            return hash;
        }
```

...which means it will throw NPE when the list contains null.

**Comment by stuart-marks** [comment]
@plevart wrote:
> But the question is how does having a separate CollSer.IMM_LIST_NULLS type prevent that from happening?

When a serialized list with IMM_LIST_NULLS is deserialized on an older JDK, it'll throw InvalidObjectException since that tag isn't valid on older JDKs. Obviously this is still an error, but it's a fail-fast approach that avoids letting nulls leak into a data structure where they might cause a problem some arbitrary time later.

> NOTE ALSO that ListN.equals(o) and ListN.hashCode() are inherited from AbstractImmutableList [...] which means they will throw NPE when the list contains null. The same goes for SubList.

Good catch! Yes, this is a problem. I'll do some rearranging here and add more test cases. Thanks for spotting this.

**Comment by plevart** [comment]
> @plevart wrote:
> 
> > But the question is how does having a separate CollSer.IMM_LIST_NULLS type prevent that from happening?
> 
> When a serialized list with IMM_LIST_NULLS is deserialized on an older JDK, it'll throw InvalidObjectException since that tag isn't valid on older JDKs. Obviously this is still an error, but it's a fail-fast approach that avoids letting nulls leak into a data structure where they might cause a problem some arbitrary time later.
> 

Yes, but that is JDK16+ vs. JDK15- and not App V1 vs. App V2 thing. If both apps run on JDK16+, there will be no exception.

What I'm trying to say is that the same problem of injecting unexpected nulls via serialization/deserialization can happen also if App V2 starts using ArrayList to construct the data structure and serialize it while App V1 deserializes it and expects non-null values only. App V1 would already have to guard against null values during deserialization in that case, because possibility of null values in deserialized data structure is nothing new for App V1.

**Comment by stuart-marks** [comment]
@plevart wrote:
> Yes, but that is JDK16+ vs. JDK15- and not App V1 vs. App V2 thing. If both apps run on JDK16+, there will be no exception.

Sure, the IMM_LIST_NULLS tag only helps with serialization compatibility across JDK releases. There are lots of ways an app can make incompatible changes to the serialized forms of its objects that we have no way of detecting.

**Comment by stuart-marks** [comment]
>> Sure, the IMM_LIST_NULLS tag only helps with serialization compatibility across JDK releases.

> I would say it goes the other way - it worsens the serialization
compatibility.

OK, I was imprecise. The IMM_LIST_NULLS tag has an effect only on serialization across JDK releases, not changes to the application's serialization format using the same JDK release, or even on many changes to the app's serialization format across JDK releases. By "helps with serialization compatibility" I meant that this new serialized form helps the general issue of serialization compatibility (really, incompatibility) by failing fast in certain cases, instead of possibly allowing polluted data to leak into the receiving application and causing some arbitrary exception later during the run.

But as you noted last, this is a different kind of object, and it has different behavior, so it needs a different encoding in the serialized form.

I'll update this PR shortly with changes to fix null handling and other issues.

File: src/java.base/share/classes/java/util/ImmutableCollections.java (hunk 6 above)
Comment (forax) [reply src/java.base/share/classes/java/util/ImmutableCollections.java:201]: you can still calls the varargs with an already created array
listFromTrustedArray(new String[] { "foo" });

I think at least an assert is missing
assert input.getClass() == Object.class;

File: src/java.base/share/classes/java/util/ImmutableCollections.java [hunk 8]
```diff
@@ -148,18 +152,101 @@ private ImmutableCollections() { }
...
+        switch (input.length) {
+            case 0:
+                return (List<E>) ImmutableCollections.EMPTY_LIST;
+            case 1:
+                return (List<E>) new List12<>(input[0]);
+            case 2:
+                return (List<E>) new List12<>(input[0], input[1]);
+            default:
+                return (List<E>) new ListN<>(input, false);
+        }
```
Comment (forax) [inline src/java.base/share/classes/java/util/ImmutableCollections.java:222]: Using a switch expression + arrow (->) here will allow you too factor the cast, even if it disappear in the generated bytecode, I think it makes the code more readable

File: src/java.base/share/classes/java/util/stream/Stream.java (hunk 4 above)
Comment (forax) [reply src/java.base/share/classes/java/util/stream/Stream.java:1196]: This implementation duplicates the array twice, once in this.toArray() and once in the constructor of ArrayList that calls toArray() on Collections.ArrayList.

I'm sure there is a better way

Comment (sergus13) [reply src/java.base/share/classes/java/util/stream/Stream.java:1196]: In `ReferencePipeline` class we have:
`
    @Override
    public final Object[] toArray() {
        return toArray(Object[]::new);
    }

    @Override
    @SuppressWarnings("unchecked")
    public final <A> A[] toArray(IntFunction<A[]> generator) {
        ...
        @SuppressWarnings("rawtypes")
        IntFunction rawGenerator = (IntFunction) generator;
        return (A[]) Nodes.flatten(evaluateToArrayNode(rawGenerator), rawGenerator)
                              .asArray(rawGenerator);
    }

`
In `Nodes` class we have:
`
    public static <T> Node<T> flatten(Node<T> node, IntFunction<T[]> generator) {
        if (node.getChildCount() > 0) {
            ... 
            **T[] array = generator.apply((int) size);**
            new ToArrayTask.OfRef<>(node, array, 0).invoke();
            return node(array);
        } else {
            return node;
        }
    }
`

It looks like it is required to implement `toList` method in a similar way in order to avoid array copy. i.e. there will be `IntFunction<List<T>> generator` which will generate 'ArrayList' with specified size and the list's `add` method will be called to add element to the list.

Comment (stuart-marks) [reply src/java.base/share/classes/java/util/stream/Stream.java:1196]: This is the default implementation in the Stream interface, which is overridden by an implementation in the ReferencePipeline class, so it will rarely be used in normal operation. The ReferencePipeline version (part of this changeset) is based on toArray() and avoids any copying. I'm thus not inclined to add new interfaces in order to support this default implementation.

As written it's true that the default implementation does perform apparently redundant copies, but we can't be assured that toArray() actually returns a freshly created array. Thus, we wrap it using Arrays.asList and then copy it using the ArrayList constructor. This is unfortunate but necessary to avoid situations where someone could hold a reference to the internal array of a List, allowing modification of a List that's supposed to be unmodifiable.

File: src/java.base/share/classes/java/util/ImmutableCollections.java (hunk 6 above)
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/ImmutableCollections.java:201]: If the parameter were of type `E...`, then a call such as `listFromTrustedArray("foo", "bar")` would result in a `String[]`. The idea is to avoid accidents such as the one that caused JDK-6260652, and not prevent deliberate passing of an array of the wrong type, though this is an internal interface so it's unlikely to occur. Still, a stronger comment and an assert might be worthwhile.

File: src/java.base/share/classes/java/util/ImmutableCollections.java (hunk 8 above)
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/ImmutableCollections.java:222]: Heh yes, I haven't gotten used to using those yet!

File: src/java.base/share/classes/java/util/stream/Stream.java (hunk 4 above)
Comment (plevart) [reply src/java.base/share/classes/java/util/stream/Stream.java:1196]: An alternative with similar performance would be to do a Stream.toArray() and then copy that array into new Object[] and then wrap that copy with listFromTrustedArrayNullsAllowed(). The difference would be in the serialization format of the resulting List and maybe also in the access performance of resulting List (no indirection via the unmodifiableList wrapper and different type for JIT to speculate about). So if we want the resulting List to behave exactly the same in both implementations of toList(), then this alternative might be preferable. WDYT?

Comment (plevart) [reply src/java.base/share/classes/java/util/stream/Stream.java:1196]: Such alternative might also be faster using intrisified array copying method (here measuring just copying overhead):

```
Benchmark              (len)  Mode  Cnt       Score      Error  Units
ToListBench.toList1       10  avgt   10      14.213 ±    0.061  ns/op
ToListBench.toList1     1000  avgt   10     541.883 ±    3.845  ns/op
ToListBench.toList1  1000000  avgt   10  753223.523 ± 4656.664  ns/op
ToListBench.toList2       10  avgt   10       8.810 ±    0.052  ns/op
ToListBench.toList2     1000  avgt   10     264.748 ±    0.807  ns/op
ToListBench.toList2  1000000  avgt   10  349518.502 ± 3242.061  ns/op
```

https://gist.github.com/plevart/974b67b65210f8dd122773f481c0a603

File: src/java.base/share/classes/java/util/ImmutableCollections.java [hunk 9]
```diff
@@ -148,18 +152,101 @@ private ImmutableCollections() { }
...
+    // ---------- List Static Factory Methods ----------
 
-    // make a copy, short-circuiting based on implementation class
+    /**
+     * Copies a collection into a new List, unless the arg is already a safe,
+     * null-prohibiting unmodifiable list, in which case the arg itself is returned.
+     * Null argument or null elements in the argument will result in NPE.
+     *
+     * @param <E> the List's element type
+     * @param input the input array
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/util/ImmutableCollections.java:163]: Rogue parameter declaration.

**Review by PaulSandoz — APPROVED** [review]
Recommend you review the problem list of this source file in IntelliJ, which shows some potential cleanups. This can be done as a follow on PR if appropriate, as can any improvements to the default implementation.

File: src/java.base/share/classes/java/util/stream/Stream.java (hunk 4 above)
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/stream/Stream.java:1196]: I don't want to change the default implementation and its specification, for a couple reasons. First, this implementation won't be used in normal operation, as it's overridden by the JDK implementation. (I expect that stream extension libraries will be updated to override it as well.) Second, I'm quite uncomfortable using an internal method from within a default implementation. Right now this method is an agreement between the unmodifiable collections implementation and the ReferencePipeline implementation, and we have complete freedom to change this agreement at any time. If this internal method were used from a default implementation, its behavior would have to be specified in `@implSpec` somehow. Certainly one could write some very vague words for the specification that allow some freedom to make changes, but this has to choose between a better specification and more implementation freedom. I don't see the value in having to make this tradeoff at all.

An extra copy can be avoided via a private agreement between ArrayList and Arrays.asList, which should probably be done in any case.
